x/iro/keeper/trade.go

```diff
@@ -0,0 +1,215 @@
+package keeper
+
+import (
+	"context"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	appparams "github.com/dymensionxyz/dymension/v3/app/params"
+	"github.com/dymensionxyz/dymension/v3/x/iro/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v15/x/txfees/types"
+)
+
+var AllocationSellLimit = math.LegacyNewDecWithPrec(999, 3) // 99.9%
+
+// Buy implements types.MsgServer.
+func (m msgServer) Buy(ctx context.Context, req *types.MsgBuy) (*types.MsgBuyResponse, error) {
+	buyer, err := sdk.AccAddressFromBech32(req.Buyer)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.Keeper.Buy(sdk.UnwrapSDKContext(ctx), req.PlanId, buyer, req.Amount, req.ExpectedOutAmount)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.MsgBuyResponse{}, nil
+}
+
+// Sell implements types.MsgServer.
+func (m msgServer) Sell(ctx context.Context, req *types.MsgSell) (*types.MsgSellResponse, error) {
+	seller, err := sdk.AccAddressFromBech32(req.Seller)
+	if err != nil {
+		return nil, err
+	}
+	err = m.Keeper.Sell(sdk.UnwrapSDKContext(ctx), req.PlanId, seller, req.Amount, req.ExpectedOutAmount)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.MsgSellResponse{}, nil
+}
+
+// Buy buys allocation with price according to the price curve
+func (k Keeper) Buy(ctx sdk.Context, planId string, buyer sdk.AccAddress, amountTokensToBuy, maxCost math.Int) error {
+	plan, found := k.GetPlan(ctx, planId)
+	if !found {
+		return types.ErrPlanNotFound
+	}
+
+	err := k.validateIROTradeable(ctx, plan, buyer.String())
+	if err != nil {
+		return err
+	}
+
+	// validate the IRO have enough tokens to sell
+	// protocol will apply max limit (99.9%) to enforce initial token liquidity
+	maxSellAmt := plan.TotalAllocation.Amount.ToLegacyDec().Mul(AllocationSellLimit).TruncateInt()
+	if plan.SoldAmt.Add(amountTokensToBuy).GT(maxSellAmt) {
+		return types.ErrInsufficientTokens
+	}
+
+	// Calculate cost for buying amountTokensToBuy over fixed price curve
+	cost := plan.BondingCurve.Cost(plan.SoldAmt, plan.SoldAmt.Add(amountTokensToBuy))
+	if !cost.IsPositive() {
```

Comment:
imo better to handle this case in the `Cost` function vs on both `buy` and `sell`